controllers/deactivation/predicate.go

```diff
@@ -25,3 +30,30 @@ func (CreateAndUpdateOnlyPredicate) Delete(_ event.DeleteEvent) bool {
 func (CreateAndUpdateOnlyPredicate) Generic(_ event.GenericEvent) bool {
 	return false
 }
+
+type GenerationOrConditionsChangedPredicate struct {
+	predicate.GenerationChangedPredicate
+}
```

Comment:
related to the other comments in the other PR, could you please explain why you want to introduce this predicate? What is the benefit of it?

Reply:
but why doing it via a predicate and not in the controller logic?
btw, this is exactly why we shouldn't touch status from different controllers that rely on updates of the status. This is why the separation of concerns matters.

Reply:
We can't do it in the controller logic because we don't have a temporal point of reference.  The deactivation time is calculated as the provisioned time of the MUR plus the number of days in the user's tier but only when the user is not in a deactivating state.  When the user is in a deactivating state then it is calculated as the last transition time of the deactivating notification sent condition plus the number of deactivating notification days.  

Before the notification is actually sent we are left in a state of limbo when we have no way of calculating the scheduled deactivation time besides a calculation based on the current time, which is what the controller currently does in this PR.  The notification may fail to send for whatever hypothetical reason of your choice, but in the end this means that the user may be stuck in this state for an extended period of time.  In this case the worst that will happen is that the user sees they have "3" days left in the dashboard for their trial, and if it continues this way it will remain at 3 days until the notification is sent successfully.

There is already a lot of complex logic surrounding the deactivation process and Alexey and I have discussed the best way to approach this issue and this solution seems to provide the best experience for the user.

Reply:
put it simply:
* if in deactivating state, set it to 3 days since the notification was created
* if not and provisioned, set it to MUR provisioned time + number of days given by the tier
* in all other cases, either set it to nil (in some error cases or when user is deprovisioned), or keep the same value

in some error cases, sure, there could be some bad numbers, but let's be practical here and keep the code simple.
As for the case you mentioned above, if the notification is not send, then you can simply control what is set to the field directly in the method `setStatusDeactivatingNotificationCreationFailed`. In this way, it's much easier to control what is going on and to follow the states of the UserSignup resource.